### PR TITLE
nmcli: add wifi support

### PR DESCRIPTION
**Pull request by goneri**
Ability to connect to a Wifi network and also to attach the connection
to a master (bond).

### CODE REVIEW

File: changelogs/fragments/2220_nmcli_wifi_support.yaml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+---
+minor_changes:
+- "nmcli - add ability to connect to a Wifi network and also to attach the connection to a master (bond)."
```
Comment (aminvakil) [inline changelogs/fragments/2220_nmcli_wifi_support.yaml:3]:
```suggestion
- "nmcli - add ability to connect to a Wifi network and also to attach the connection to a master (bond) (https://github.com/ansible-collections/community.general/pull/2220)."
```

File: plugins/modules/net_tools/nmcli.py [hunk 2]
```diff
@@ -278,7 +278,14 @@
             - The trust level of the connection.
             - When updating this property on a currently activated connection, the change takes effect immediately.
        type: str
-       version_added: 2.0.0
+    wifi_sec:
+       description:
+            - The security configuration of the Wifi connection
```
Comment (aminvakil) [inline plugins/modules/net_tools/nmcli.py:283]: Could you please add some examples for this option? Or maybe add this link for further reference (https://developer.gnome.org/NetworkManager/stable/settings-802-11-wireless-security.html)?
Overally I guess it would be good to improve documentation for this option on how to use.
```suggestion
            - The security configuration of the Wifi connection.
```

File: plugins/modules/net_tools/nmcli.py [hunk 3]
```diff
@@ -278,7 +278,14 @@
...
+       type: dict
+    ssid:
+       description:
+            - Name of the Wireless router or the access points.
```
Comment (aminvakil) [inline plugins/modules/net_tools/nmcli.py:287]:
```suggestion
            - Name of the wireless router or the access point.
            - Required if I(type=wifi).
```

**Review by aminvakil — COMMENTED** [review]
This is pretty useful, thanks for adding this!

File: plugins/modules/net_tools/nmcli.py [hunk 4]
```diff
@@ -278,7 +278,14 @@
             - The trust level of the connection.
             - When updating this property on a currently activated connection, the change takes effect immediately.
        type: str
-       version_added: 2.0.0
```
Comment (felixfontein) [inline plugins/modules/net_tools/nmcli.py:281]: This line needs to stay, and the new options need `version_added: 2.5.0`.

**Comment by goneri** [comment]
Thank you both for the super quick feedbaacks.

File: plugins/modules/net_tools/nmcli.py [hunk 5]
```diff
@@ -279,6 +279,20 @@
             - When updating this property on a currently activated connection, the change takes effect immediately.
        type: str
        version_added: 2.0.0
+    wifi_sec:
+       description:
+            - 'The security configuration of the Wifi connection. The valid attributes are listed on:'
+            - 'https://developer.gnome.org/NetworkManager/stable/settings-802-11-wireless-security.html'
```
Comment (felixfontein) [inline plugins/modules/net_tools/nmcli.py:285]:
```suggestion
            - 'U(https://developer.gnome.org/NetworkManager/stable/settings-802-11-wireless-security.html)'
```

File: plugins/modules/net_tools/nmcli.py [hunk 6]
```diff
@@ -1141,8 +1184,11 @@ def main():
             ip_tunnel_dev=dict(type='str'),
             ip_tunnel_local=dict(type='str'),
             ip_tunnel_remote=dict(type='str'),
+            ssid=dict(type='str'),
+            wifi_sec=dict(type='dict'),
```
Comment (felixfontein) [inline plugins/modules/net_tools/nmcli.py:1188]: Since this option contains credentials:
```suggestion
            wifi_sec=dict(type='dict', no_log=True),
```

File: plugins/modules/net_tools/nmcli.py [hunk 7]
```diff
@@ -279,6 +279,20 @@
             - When updating this property on a currently activated connection, the change takes effect immediately.
        type: str
        version_added: 2.0.0
+    wifi_sec:
+       description:
+            - 'The security configuration of the Wifi connection. The valid attributes are listed on:'
+            - 'https://developer.gnome.org/NetworkManager/stable/settings-802-11-wireless-security.html'
+            - 'For instance to use common WPA-PSK auth with a password:'
+            - '  - key-mgmt: wpa-psk'
+            - '  - psk: my_password'
```
Comment (felixfontein) [inline plugins/modules/net_tools/nmcli.py:288]: This should be a dictionary, not a list of dictionaries, right? It's probably better to format it like this:
```suggestion
            - 'C({key-mgmt: wpa-psk, psk: my_password})'
```

File: plugins/modules/net_tools/nmcli.py [hunk 8]
```diff
@@ -695,7 +725,7 @@ def connection_options(self, detect_change=False):
         }
 
         # IP address options.
-        if self.ip_conn_type:
+        if self.ip_conn_type and not self.master:
```
Comment (felixfontein) [inline plugins/modules/net_tools/nmcli.py:728]: This change looks unrelated to this PR.

Comment (goneri) [reply plugins/modules/net_tools/nmcli.py:728]: Ok, I create another PR.

File: plugins/modules/net_tools/nmcli.py [hunk 9]
```diff
@@ -582,6 +595,20 @@
...
+    type: wifi
+    conn_name: Brittany
+    ifname: wlp4s0
+    ssid: Brittany
+    wifi_sec:
+      key-mgmt: wpa-psk
+      psk: my_password
+    autoconnect: true
+    state: present
+  no_log: true
```
Comment (felixfontein) [inline plugins/modules/net_tools/nmcli.py:610]: Is this still needed?

Comment (goneri) [reply plugins/modules/net_tools/nmcli.py:610]: I think we should keep it, it's a very standard use-case. I had to actually refrain myself from changing the rest of the examples to something more standard.

Comment (felixfontein) [reply plugins/modules/net_tools/nmcli.py:610]: I mean the `no_log: true` (in case you meant the whole example).

Comment (goneri) [reply plugins/modules/net_tools/nmcli.py:610]: OH! Yes, I can remove the line.

File: changelogs/fragments/2220_nmcli_wifi_support.yaml [hunk 10]
```diff
@@ -0,0 +1,3 @@
+---
+minor_changes:
+- "nmcli - add ability to connect to a Wifi network and also to attach the connection to a master (bond) (https://github.com/ansible-collections/community.general/pull/2220)."
```
Comment (aminvakil) [inline changelogs/fragments/2220_nmcli_wifi_support.yaml:3]: This change makes it more readable and easier to understand IMHO, but it's fine as it is now as well.
```suggestion
- "nmcli - add ability to connect to a Wifi network and attach it to a master (bond) (https://github.com/ansible-collections/community.general/pull/2220)."
```

**Review by aminvakil — APPROVED** [review]
LGTM

File: plugins/modules/net_tools/nmcli.py [hunk 11]
```diff
@@ -279,6 +279,19 @@
             - When updating this property on a currently activated connection, the change takes effect immediately.
        type: str
        version_added: 2.0.0
+    wifi_sec:
+       description:
+            - 'The security configuration of the Wifi connection. The valid attributes are listed on:'
+            - 'U(https://developer.gnome.org/NetworkManager/stable/settings-802-11-wireless-security.html)'
+            - 'For instance to use common WPA-PSK auth with a password:'
+            - '- C({key-mgmt: wpa-psk, psk: my_password})'
+       type: dict
```
Comment (russoz) [inline plugins/modules/net_tools/nmcli.py:288]: It might be interesting to actually add the suboptions here, and mark the wifi passwords as `no_log`.

File: plugins/modules/net_tools/nmcli.py [hunk 12]
```diff
@@ -919,6 +952,13 @@ def connection_update(self, nmcli_command):
         else:
             ifname = self.ifname
 
+        if self.type == "wifi":
+            cmd.append('ssid')
+            cmd.append(self.ssid)
+            if self.wifi_sec:
+                for name, value in self.wifi_sec.items():
+                    cmd += ['wifi-sec.%s' % name, value]
```
Comment (russoz) [inline plugins/modules/net_tools/nmcli.py:960]: a bit of a nit-picking here but
```suggestion
                    cmd.extend(['wifi-sec.%s' % name, value])
```
would be slightly more efficient.

File: plugins/modules/net_tools/nmcli.py (hunk 11 above)
Comment (goneri) [reply plugins/modules/net_tools/nmcli.py:288]: If we list the `suboptions` we create a strong dependency with one specific NetworkManager version. I prefer to offer as much flexibility as possible.

File: changelogs/fragments/2220_nmcli_wifi_support.yaml (hunk 10 above)
Comment (goneri) [reply changelogs/fragments/2220_nmcli_wifi_support.yaml:3]: Indeed. Thanks.

File: plugins/modules/net_tools/nmcli.py [hunk 13]
```diff
@@ -279,6 +279,19 @@
...
        type: str
        version_added: 2.0.0
+    wifi_sec:
+       description:
+            - 'The security configuration of the Wifi connection. The valid attributes are listed on:'
+            - 'U(https://developer.gnome.org/NetworkManager/stable/settings-802-11-wireless-security.html)'
+            - 'For instance to use common WPA-PSK auth with a password:'
+            - '- C({key-mgmt: wpa-psk, psk: my_password})'
+       type: dict
+       version_added: 2.5.0
```
Comment (aminvakil) [inline plugins/modules/net_tools/nmcli.py:289]:
```suggestion
       version_added: 3.0.0
```

File: plugins/modules/net_tools/nmcli.py [hunk 14]
```diff
@@ -279,6 +279,19 @@
...
+    ssid:
+       description:
+            - Name of the Wireless router or the access point.
+       type: str
+       version_added: 2.5.0
```
Comment (aminvakil) [inline plugins/modules/net_tools/nmcli.py:294]:
```suggestion
       version_added: 3.0.0
```

**Review by aminvakil — COMMENTED** [review]
Unfortunately this PR missed `2.5.0` release too :(

File: plugins/modules/net_tools/nmcli.py (hunk 12 above)
Comment (goneri) [reply plugins/modules/net_tools/nmcli.py:960]: Thanks! The difference is actually huge:
```
[goneri@t580-gone ~]$ python3 -m timeit -s 'a = []' 'for n in range(1000): a += [n]'
5000 loops, best of 5: 82.9 usec per loop
[goneri@t580-gone ~]$ python3 -m timeit -s 'a = []' 'for n in range(1000): a.append(n)'
5000 loops, best of 5: 53.8 usec per loop
```

Comment (goneri) [reply plugins/modules/net_tools/nmcli.py:960]: Hum my bad, with the `extend()`, it's actually slow:

```
[goneri@t580-gone ~]$ python3 -m timeit -s 'a = []' 'for n in range(1000): a.extend([n])'
5000 loops, best of 5: 94.5 usec per loop
```

Comment (goneri) [reply plugins/modules/net_tools/nmcli.py:960]: It's also the case with py27:

```
[goneri@t580-gone ~]$ python2 -m timeit -s 'a = []' 'for n in range(1000): a.extend([n])'
10000 loops, best of 3: 84.6 usec per loop
[goneri@t580-gone ~]$ python2 -m timeit -s 'a = []' 'for n in range(1000): a += [n]'
10000 loops, best of 3: 70.8 usec per loop
```

Comment (felixfontein) [reply plugins/modules/net_tools/nmcli.py:960]: Interesting, I would have expected `extend()` to not be slower. If you add the string formatting though, the relative difference is pretty small though.

(The Python 2 performance is something I would ignore here, except if this would be something that's noticable to users - which this tiny difference clearly isn't :) )

One could argue that it's better to use `extend()` for consistency though, since it's more similar to `append()` used everywhere else. I'm not completely convinced by that. I don't care enough to ask for this to change, or ask for this to not change ;) @russoz what do you think?

File: plugins/modules/net_tools/nmcli.py (hunk 11 above)
Comment (felixfontein) [reply plugins/modules/net_tools/nmcli.py:288]: It's kind of a PITA that this is dependent on the NM version.

Maybe the module should provide one consistent interface and convert it to whatever the nmcli version in use requires?

Comment (goneri) [reply plugins/modules/net_tools/nmcli.py:288]: Why? This would complexify the logic and the module would still require a manual to support new keys.

Comment (felixfontein) [reply plugins/modules/net_tools/nmcli.py:288]: Mainly because it is easier for users, who can use the same task for different nm versions without having to adjust the parameters depending on the nm version. Also it allows to have a better control on the suboptions, so that only the passphrase itself is marked as `no_log`, while the other values are not. Right now, if any non-secret value in this `dict` also appears anywhere else in the module output, it will be censored. Which can potentially make the output pretty unhelpful.

File: plugins/modules/net_tools/nmcli.py (hunk 12 above)
Comment (russoz) [reply plugins/modules/net_tools/nmcli.py:960]: I think that for fairness we should test `extend` with lists longer than 1 element.

That said, for the sake of this PR I really don't think we should press the point. I would also prefer `append` or `extend` for consistency, but if @goneri is happy with `+` I say let's go with it.

**Comment by russoz** [comment]
Pending the discussion on extending lists, LGTM.

File: plugins/modules/net_tools/nmcli.py (hunk 12 above)
Comment (felixfontein) [reply plugins/modules/net_tools/nmcli.py:960]: @russoz I tested it with two elements, `+=` is still faster.

**Comment by felixfontein** [comment]
@goneri thanks for contributing this!
@aminvakil @russoz thanks for reviewing!
